File: day20.c
#include <stdio.h>
#define MIN(a,b) (((a)<(b))?(a):(b))
#define MAX(a,b) (((a)>(b))?(a):(b))

int N=0;
int abs(int x){return x>0?x:-x;}
int hash(int x,int y){return N*y+x;}
int main(void){
    FILE *f=fopen("data/input20.txt","r");
    char G[150][150];
    int dist[150*150]={0};
    while(fgets(G[N],150,f)&&G[N][0]!='\n'){N++;}
    int sx,sy,ex,ey;
    for(int y=0;y<N;y++){
        for(int x=0;x<N;x++){
            if(G[y][x]=='S'){sx=x;sy=y;}
            if(G[y][x]=='E'){ex=x;ey=y;}
        }
    }

    int x=sx,y=sy,d=0;
    while(1){
        if(G[y][x]=='#')continue;
        dist[hash(x,y)]=d+1;
        if(x==ex && y==ey)break;
        d++;
        if(x>0 && G[y][x-1]!='#' && dist[hash(x-1,y)]==0)x--;
        else if(x<N-1 && G[y][x+1]!='#' && dist[hash(x+1,y)]==0)x++;
        else if(y>0 && G[y-1][x]!='#' && dist[hash(x,y-1)]==0)y--;
        else if(y<N-1 && G[y+1][x]!='#' && dist[hash(x,y+1)]==0)y++;
    }

    int part1=0,part2=0;
    int D=dist[hash(ex,ey)]-1;
    for(int y=1; y<N-1; y++){
        for(int x=1; x<N-1; x++){
            if(G[y][x]=='#')continue;
            int d1=dist[hash(x,y)]-1;
            for(int yy=MAX(1,y-20);yy<MIN(N-1,y+21);yy++){
                for(int xx=MAX(1,x-20);xx<MIN(N-1,x+21);xx++){
                    if(G[yy][xx]=='#')continue;
                    int d2=dist[hash(ex,ey)]-dist[hash(xx,yy)];
                    int d3=abs(x-xx)+abs(y-yy);
                    if(d3<=20){
                        int d=d1+d2+d3;
                        if(d3==2 && d<=D-100)part1++;
                        if(d<=D-100)part2++;
                    }
                }
            }
        }        
    }
    printf("%d %d\n",part1,part2);
    return 0;
}

File: day20.lua
local part1,part2=0,0
local G={}
local sx,sy,W,H=0,0,0,0
for line in io.lines("data/input20.txt") do
    if #line>0 then
        H=H+1
        local row={}
        for i=1,#line do
            local ch=string.sub(line,i,i)
            table.insert(row,ch)
            if ch=="S" then sx,sy=i,H end
            if ch=="E" then ex,ey=i,H end
        end
        table.insert(G,row)
    end
end
W=#G[1]

function hash(x,y) return 150*y+x end

function distMap(sx,sy,ex,ey)
    local dist={}
    local x,y=sx,sy 
    local d=0  
    while true do
        dist[hash(x,y)]=d
        if x==ex and y==ey then return dist end
        d=d+1
        if x>1 and G[y][x-1]~="#" and not dist[hash(x-1,y)] then x=x-1
        elseif x<W and G[y][x+1]~="#" and not dist[hash(x+1,y)] then x=x+1
        elseif y>1 and G[y-1][x]~="#" and not dist[hash(x,y-1)] then y=y-1
        elseif y<H and G[y+1][x]~="#" and not dist[hash(x,y+1)] then y=y+1; end 
    end
end

local dist1=distMap(sx,sy,ex,ey)
local dist2=distMap(ex,ey,sx,sy)
local D=dist1[hash(ex,ey)]
for y=2,H-1 do
    for x=2,W-1 do
        if G[y][x]=="#" then goto skip_outer end
            local d1=dist1[hash(x,y)]
            for yy=math.max(2,y-20),math.min(H-1,y+20) do
                for xx=math.max(2,x-20),math.min(W-1,x+20) do             
                    if G[yy][xx]=="#" then goto skip_inner end
                        local d2=dist2[hash(xx,yy)]
                        local d3=math.abs(x-xx)+math.abs(y-yy)
                        if d3<=20 then
                            local d=d1+d2+d3
                            if d3==2 and d<=D-100 then part1=part1+1 end
                            if d<=D-100 then part2=part2+1 end
                        end
                    ::skip_inner::
                end
            end
        ::skip_outer::
    end
end
print(part1,part2)